services/motion/builtin/moveonglobe.go

```diff
@@ -0,0 +1,314 @@
+package builtin
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"sync"
+	"time"
+
+	geo "github.com/kellydunn/golang-geo"
+	"github.com/pkg/errors"
+	goutils "go.viam.com/utils"
+
+	"go.viam.com/rdk/components/base"
+	"go.viam.com/rdk/components/base/kinematicbase"
+	"go.viam.com/rdk/components/movementsensor"
+	"go.viam.com/rdk/motionplan"
+	"go.viam.com/rdk/referenceframe"
+	"go.viam.com/rdk/resource"
+	"go.viam.com/rdk/services/motion"
+	"go.viam.com/rdk/spatialmath"
+)
+
+type replanResponse struct {
+	err    error
+	replan bool
+}
+
+type executeResponse struct {
+	err     error
+	success bool
+}
+
+type planSession struct {
+	plan              motionplan.Plan
+	executionChan     chan executeResponse
+	positionChan      chan replanResponse
+	obsticleChan      chan replanResponse
+	cancelCtx         context.Context
+	cancelFn          context.CancelFunc
+	backgroundWorkers *sync.WaitGroup
+}
+
+func plan(
+	ctx context.Context,
+	planRequest *motionplan.PlanRequest,
+	kb kinematicbase.KinematicBase,
+	componentName string,
+) (motionplan.Plan, error) {
+	inputs, err := kb.CurrentInputs(ctx)
+	if err != nil {
+		return make(motionplan.Plan, 0), err
+	}
+	// TODO: this is really hacky and we should figure out a better place to store this information
+	if len(kb.Kinematics().DoF()) == 2 {
+		inputs = inputs[:2]
+	}
+	planRequest.StartConfiguration = map[string][]referenceframe.Input{componentName: inputs}
+
+	return motionplan.PlanMotion(ctx, planRequest)
+}
+
+func startPollingForReplan(
+	ctx context.Context,
+	period time.Duration,
+	responseChan chan replanResponse,
+	backgroundWorkers *sync.WaitGroup,
+	fn func(context.Context,
+	) replanResponse,
+) {
+	backgroundWorkers.Add(1)
+	goutils.ManagedGo(func() {
+		ticker := time.NewTicker(period)
+		defer ticker.Stop()
+		for {
+			// this ensures that if the context is cancelled we always
+			// return early at the top of the loop
+			if err := ctx.Err(); err != nil {
```

Comment:
This fixes a bug where the select statement might not go down the ctx.Done() path if there are messages on the other channels. This change ensures that we always terminate the loop if the context has been cancelled.